# Atelier – Préparation de Données Tabulaires

**Contexte :** une entreprise exploite plusieurs bâtiments intelligents équipés de capteurs IoT
(température, humidité, qualité de l'air, consommation énergétique, occupants, type de bâtiment,
mode de fonctionnement, état de la climatisation).

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

ModuleNotFoundError: No module named 'seaborn'

## Partie 1 – Explorer les données

### 1) Charger les données CSV

Le fichier brut est chargé depuis le sous-dossier `data/`. 


In [ ]:
df=pd.read_csv('../data/smart_building_raw.csv')


,id_mesure,date,batiment,type_batiment,zone,temperature,humidite,co2,occupation,consommation_kwh,mode_climatisation,etat_systeme,jour_semaine,alerte
0,1174,2025-02-13 06:00:00,B8,Entrepôt,A,23.0,40.0,851.0,26.0,138.4,Eco,Normal,Jeudi,Non
1,1275,2025-03-10 12:00:00,B7,Bureau,D,28.0,69.0,538.0,0.0,170.2,Eco,Normal,Lundi,Non
2,1493,2025-05-04 00:00:00,B5,Centre commercial,B,19.5,52.3,1198.0,64.0,149.8,Normal,Normal,Dimanche,Oui
3,1073,2025-01-19 00:00:00,B6,Université,D,20.0,58.4,1014.0,30.0,141.6,Normal,Normal,Dimanche,Non
4,1454,2025-04-24 06:00:00,B7,Bureau,D,26.8,70.5,628.0,42.0,228.3,Normal,Normal,Jeudi,Non


### 2) Afficher les premières lignes du dataset



In [4]:
df.head()

,id_mesure,date,batiment,type_batiment,zone,temperature,humidite,co2,occupation,consommation_kwh,mode_climatisation,etat_systeme,jour_semaine,alerte
0,1174,2025-02-13 06:00:00,B8,Entrepôt,A,23.0,40.0,851.0,26.0,138.4,Eco,Normal,Jeudi,Non
1,1275,2025-03-10 12:00:00,B7,Bureau,D,28.0,69.0,538.0,0.0,170.2,Eco,Normal,Lundi,Non
2,1493,2025-05-04 00:00:00,B5,Centre commercial,B,19.5,52.3,1198.0,64.0,149.8,Normal,Normal,Dimanche,Oui
3,1073,2025-01-19 00:00:00,B6,Université,D,20.0,58.4,1014.0,30.0,141.6,Normal,Normal,Dimanche,Non
4,1454,2025-04-24 06:00:00,B7,Bureau,D,26.8,70.5,628.0,42.0,228.3,Normal,Normal,Jeudi,Non


### 3) Afficher les dernières lignes du dataset



In [5]:
df.tail()

,id_mesure,date,batiment,type_batiment,zone,temperature,humidite,co2,occupation,consommation_kwh,mode_climatisation,etat_systeme,jour_semaine,alerte
502,1107,2025-01-27 12:00:00,B2,École,C,29.8,66.5,778.0,48.0,243.5,Normal,Normal,Lundi,Non
503,1271,2025-03-09 12:00:00,B3,Hôpital,B,21.1,66.3,3900.0,53.0,145.8,Eco,Alerte,Dimanche,Oui
504,1349,2025-03-29 00:00:00,B5,Centre commercial,B,21.6,61.9,959.0,33.0,100.2,Eco,Normal,Samedi,Non
505,1436,2025-04-19 18:00:00,B7,Bureau,C,21.2,54.5,1066.0,38.0,150.9,Boost,Normal,Samedi,Oui
506,1103,2025-01-26 12:00:00,B2,École,D,NaN,64.8,584.0,29.0,163.4,Normal,Panne,Dimanche,Non


### 4) Combien d'observations contient le dataset ?

Le nombre d'observations correspond au nombre de lignes du DataFrame.

In [6]:
df.shape[0]

507

### 5) Combien de variables possède le dataset ?

Le nombre de variables correspond au nombre de colonnes.

In [7]:
df.shape[1]


14

### 6) Identifier les variables numériques

Une variable est numérique lorsqu'elle représente une quantité mesurable sur laquelle
les opérations arithmétiques ont un sens (moyenne, écart, somme).


In [13]:
variables_numeriques = df.select_dtypes(include=[np.number]).columns.tolist()

print("Variables numériques détectées par pandas :")
for col in variables_numeriques:
    print(f" {col} ({df[col].dtype})")



Variables numériques détectées par pandas :
 id_mesure (int64)
 temperature (float64)
 humidite (float64)
 co2 (float64)
 occupation (float64)
 consommation_kwh (float64)


**Réponse :** Les variables numériques sont : `temperature`, `humidite`, `co2`, `occupation`,
`consommation_kwh`.

La colonne `id_mesure` est également numérique (`int64`) mais il ne s'agit pas d'une mesure :
c'est un identifiant, donc elle est exclue de cette liste au sens métier.

### 7) Identifier les variables catégorielles

Une variable est catégorielle lorsqu'elle prend un nombre restreint de modalités
qui désignent des groupes, sans relation d'ordre arithmétique.




In [14]:
variables_categorielles = df.select_dtypes(include=["object", "category", "bool"]).columns.tolist()

print("Variables catégorielles (type texte) et leur cardinalité :\n")
for col in variables_categorielles:
    modalites = df[col].dropna().unique()
    apercu = ", ".join(map(str, modalites[:8]))
    suite = " ..." if len(modalites) > 8 else ""
    print(f"  - {col:<25} {len(modalites):>3} modalités : {apercu}{suite}")


Variables catégorielles (type texte) et leur cardinalité :

  - date                      500 modalités : 2025-02-13 06:00:00, 2025-03-10 12:00:00, 2025-05-04 00:00:00, 2025-01-19 00:00:00, 2025-04-24 06:00:00, 2025-01-20 00:00:00, 2025-03-21 00:00:00, 2025-02-05 00:00:00 ...
  - batiment                    8 modalités : B8, B7, B5, B6, B3, B4, B2, B1
  - type_batiment              16 modalités : Entrepôt, Bureau, Centre commercial, Université, Hôpital, École, ÉCOLE, ecole ...
  - zone                        4 modalités : A, D, B, C
  - mode_climatisation          7 modalités : Eco, Normal, Boost, normal, BOOST, normale, Normal 
  - etat_systeme                3 modalités : Normal, Alerte, Panne
  - jour_semaine                7 modalités : Jeudi, Lundi, Dimanche, Vendredi, Mercredi, Mardi, Samedi
  - alerte                      2 modalités : Non, Oui


C:\Users\other\AppData\Local\Temp\ipykernel_7404\128847611.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  variables_categorielles = df.select_dtypes(include=["object", "category", "bool"]).columns.tolist()


**Réponse :** Les variables catégorielles sont : `batiment`, `type_batiment`, `zone`,
`mode_climatisation`, `etat_systeme`, `jour_semaine`, `alerte`.


La colonne `date` ressort également dans une détection par type texte (elle a 500 modalités
distinctes), mais elle est **exclue** de la liste des variables catégorielles : il ne s'agit
pas d'une catégorie mais d'une date, encore mal typée (`object` au lieu de `datetime`). Elle
sera traité.

### 8) Identifier les dates


In [22]:

df.info()

<class 'pandas.DataFrame'>
RangeIndex: 507 entries, 0 to 506
Data columns (total 14 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   id_mesure           507 non-null    int64  
 1   date                507 non-null    str    
 2   batiment            507 non-null    str    
 3   type_batiment       503 non-null    str    
 4   zone                507 non-null    str    
 5   temperature         495 non-null    float64
 6   humidite            496 non-null    float64
 7   co2                 500 non-null    float64
 8   occupation          501 non-null    float64
 9   consommation_kwh    502 non-null    float64
 10  mode_climatisation  502 non-null    str    
 11  etat_systeme        507 non-null    str    
 12  jour_semaine        502 non-null    str    
 13  alerte              507 non-null    str    
dtypes: float64(5), int64(1), str(8)
memory usage: 55.6 KB


**Réponse :** La colonne `date` représente la date de mesure. Elle est actuellement de type
`str` (texte), et non de type date : elle devra être convertie plus tard (`pd.to_datetime`)
avant toute analyse chronologique ou tri temporel.

### 9) Identifier les identifiants



In [30]:
print("Valeurs uniques :", df["id_mesure"].nunique())
print("Nombre de lignes :", len(df))
print("Valeurs manquantes :", df["id_mesure"].isna().sum())

Valeurs uniques : 500
Nombre de lignes : 507
Valeurs manquantes : 0


**Réponse :** La colonne `id_mesure` (`int64`, 507 valeurs non nulles) est un identifiant :
elle numérote chaque mesure de façon unique et ne porte aucune information mesurable sur le
bâtiment ou l'environnement. Elle sera exclue des variables explicatives X pour éviter toute
fuite de données.

### 10) Statistiques descriptives

`describe()` fournit d'un coup la moyenne, l'écart-type, le minimum, le maximum et les
quartiles (Q1, médiane, Q3) de chaque variable numérique.


In [46]:
stats = df[variables_numeriques].describe().T

stats = stats.rename(columns={
    "mean": "moyenne",
    "50%": "mediane",
    "min": "minimum",
    "max": "maximum",
    "std": "ecart_type",
    "25%": "Q1",
    "75%": "Q3"
})

stats.round(2)

,count,moyenne,ecart_type,minimum,Q1,mediane,Q3,maximum
id_mesure,507.0,1251.11,144.78,1001.0,1125.50,1252.00,1376.50,1500.0
temperature,495.0,24.15,7.42,-30.0,21.60,24.00,26.00,96.0
humidite,496.0,57.86,16.03,-12.0,49.28,57.55,65.75,160.0
co2,500.0,844.15,582.18,89.0,623.75,787.50,952.00,6000.0
occupation,501.0,44.85,24.95,-20.0,27.00,46.00,61.00,116.0
consommation_kwh,502.0,169.07,53.16,-100.0,136.88,169.80,202.98,336.2


### 11) Y a-t-il des variables potentiellement problématiques ?


In [48]:
df.isna().sum()

id_mesure              0
date                   0
batiment               0
type_batiment          4
zone                   0
temperature           12
humidite              11
co2                    7
occupation             6
consommation_kwh       5
mode_climatisation     5
etat_systeme           0
jour_semaine           5
alerte                 0
dtype: int64

**Réponse :** Oui, plusieurs variables sont potentiellement problématiques :

- `temperature`, `humidite`, `co2`, `occupation`, `consommation_kwh`, `type_batiment`,
  `mode_climatisation`, `jour_semaine` contiennent des valeurs manquantes (voir tableau ci-dessus).
- `date` est de type texte (`object`) alors qu'elle représente une date : elle devra être
  convertie avec `pd.to_datetime`.
- `id_mesure` est un identifiant unique par ligne : il n'apporte aucune information
  prédictive et devra être exclu des variables explicatives X.

### 12) Rechercher les données incohérentes


In [49]:
print("a) Humidité < 0 :", (df["humidite"] < 0).sum())
print("b) Humidité > 100 :", (df["humidite"] > 100).sum())

a) Humidité < 0 : 3
b) Humidité > 100 : 7
